lib/sles4sap/console_redirection.pm

```diff
@@ -0,0 +1,174 @@
+
+=head2 handle_login_prompt
+
+    handle_login_prompt();
+
+B<ssh_user> Login user
+Detects if login prompt appears and types the password.
+
+=cut
+
+sub handle_login_prompt {
+    my $pwd = get_var('_SECRET_SUT_PASSWORD', $testapi::password);
+    set_serial_term_prompt();
+    my $serial_response = wait_serial(qr/Password:\s*$|:~/i);
+    # Handle password prompt if it appears
+    if (grep /Password:\s*$/, $serial_response) {
+        type_password $pwd;
+        send_key 'ret';
+    }
+    croak 'Failed to setup serial console' unless grep /:~/, $serial_response;
+}
+
+=head2 redirection_init
+
+    redirection_init();
+
+Do preparation before redirecting console. Gets base VM id and initial setup.
+Needs to be done only once.
+
+=cut
+
+sub redirection_init {
+    # This should get worker VM id before any redirection happening
+    # ID serves as identification of the 'base' VM to return to.
+    set_var('WORKER_VM_ID', script_output 'cat /etc/machine-id');
+}
+
+=head2 set_serial_term_prompt
+
+    set_serial_term_prompt();
+
+Detects if login prompt appears and types the password.
+
+=cut
+
+sub set_serial_term_prompt {
+    $testapi::distri->{serial_term_prompt} = '';    # reset previous prompt first
+    my $current_user = script_output('whoami');
+    croak 'SSH user not defined and/or "whoami" script failed.' unless defined($current_user);
+    $testapi::distri->{serial_term_prompt} = ($current_user eq 'root' ? '# ' : '> ');
+}
+
+=head2 connect_target_to_serial
+
+    connect_target_to_serial([ssh_user=>ssh_user, target_ip=>$target_ip]);
+
+B<ssh_user> Login user - can be alternatively defined by OpenQA parameter REDIRECT_TARGET_USER
+B<target_ip> Target host IP - can be alternatively defined by OpenQA parameter REDIRECT_TARGET_IP
+Detects if login prompt appears and types the password.
+
+=cut
+
+sub connect_target_to_serial {
+    my (%args) = @_;
+    my $redirect_ip = $args{'target_ip'} // get_required_var('REDIRECT_TARGET_IP');
+    my $ssh_user = $args{'ssh_user'} // get_required_var('REDIRECT_TARGET_USER');
+
+    croak 'Missing "ssh_user" argument' unless $ssh_user;
+    croak 'Missing "redirect_ip" argument' unless $redirect_ip;
+    croak "OpenQA variable WORKER_VM_ID undefined. Run 'redirection_init()' first" unless get_var('WORKER_VM_ID');
+    croak "IP address '$redirect_ip' is not valid." unless verify_ip($redirect_ip);
+    croak 'Global variable "$searialdev" undefined' unless $serialdev;
+    croak "Console is already redirected to:" . script_output('hostname') if check_serial_redirection();
+
+    enter_cmd "ssh $ssh_opt $ssh_user\@$redirect_ip 2>&1 | tee -a /dev/$serialdev";
+    handle_login_prompt($ssh_user);
+    record_info('Redirect ON', "Serial redirection established to: $redirect_ip");
+}
+
+=head2 disconnect_target_from_serial
+
+    disconnect_target_from_serial($worker_id);
+
+Disconnects target from serial console by typing 'exit' command until host IP matches IP addr of the worker VM.
+
```

Comment:
could you add something about the `worker_id` argument? Why it is an argument with fallback to a setting (and a special setting, the one the lib configure its self)?

Reply:
Sorry, it was described improperly. 
WORKER_VM_ID is the default value set by redirect_init() and $worker_id is the option to override this value. 
- description added
- changed to keyword argument $args{worker_machine_id}